### DISCUSSION TITLE
Allow ordering/sorting of categorical x-axis values by another variable

### DISCUSSION THREAD

**Discussion by sfirke** [apache/superset#30575] [General]
**PROBLEM**

Users need to be able to sort categorical x-axis variables by something other than alphabetical or total/max/min/avg.  For instance Sun-Mon-Tue-Wed-Thu-Fri-Sat order, or Jan-Feb-Mar-Apr, or Bad-Fair-Good-Great, etc.  But Bar Chart v2 does not support this.

Take this chart (from master branch as of Sep 2024):
![image](https://github.com/user-attachments/assets/4d84e9b4-5680-4259-af83-f239e93ae937)

The x-axis is alphabetically sorted and there is no way currently to make it the desired order Sun-Mon-Tue ...

**PROPOSED SOLUTION**
Allow the x-axis sort by to use another metric, like what `Query sort by` already does:
![image](https://github.com/user-attachments/assets/20c95c9d-c47b-48cb-8049-b993ad2383b5)

Then users can create a numeric variable corresponding to their arbitrary desired order (in my case `Day_Numeric`) and control ordering that way.

**WORKAROUNDS IN THE MEANTIME**
Community members suggest appending whitespace or numerals to the beginning of the categorical variable and then using alphabetical order.  The cleanest version of this I've seen is from @marianysilva who [shows how to append invisible whitespace characters to the beginning](https://github.com/apache/superset/issues/21425#issuecomment-2403128905).

**RELATED**
This emerges from https://github.com/apache/superset/issues/21425#issuecomment-2405168584, which is originally an issue about Bar Chart v1 (legacy).  But the best solution is to support this sorting in Bar Chart v2.  

Another user asks for the same thing for the Heatmap chart in #30371.  I suspect the bar chart gets higher use and should be the priority but maybe the same fix could apply to all relevant ECharts.

**Comment by goldjee** [comment]
If I may add, this is a much needed feature not only for bar charts, but rather for each and every chart type where categorical data is supported. Some native control over ordering would be really nice. Playing with Unicode escape sequences just to achieve that looks like a forbidden magic as it renders the data dirty.

**Reply by kakoni** [reply]
I believe this is linked to this echart issue; https://github.com/apache/echarts/issues/18271

**Comment by michofwood** [comment]
This should be done at the dataset level so that for a column, it is sorted by another one. That would apply to filters and projections and anywhere a column values are exposed.
All BI tools have that intrinsict feature built in at the Data model stage.
